src/access/AccessRestriction.cpp

```diff
@@ -0,0 +1,280 @@
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void AccessRestriction::AddListener(EntryListener & listener)
+{
+    if (mListeners == nullptr)
+    {
+        mListeners     = &listener;
+        listener.mNext = nullptr;
+        return;
+    }
+
+    for (EntryListener * l = mListeners; /**/; l = l->mNext)
+    {
+        if (l == &listener)
+        {
+            return;
+        }
+
+        if (l->mNext == nullptr)
+        {
+            l->mNext       = &listener;
+            listener.mNext = nullptr;
+            return;
+        }
+    }
+}
+
+void AccessRestriction::RemoveListener(EntryListener & listener)
+{
+    if (mListeners == &listener)
+    {
+        mListeners     = listener.mNext;
+        listener.mNext = nullptr;
+        return;
+    }
+
+    for (EntryListener * l = mListeners; l != nullptr; l = l->mNext)
+    {
+        if (l->mNext == &listener)
+        {
+            l->mNext       = listener.mNext;
+            listener.mNext = nullptr;
+            return;
+        }
+    }
+}
+
+SharedPtr<AccessRestriction::Entry> AccessRestriction::GetEntry(FabricIndex fabricIndex, size_t index)
+{
+    if (mFabricEntries.find(fabricIndex) != mFabricEntries.end() && index < mFabricEntries[fabricIndex].size())
+    {
+        return mFabricEntries[fabricIndex][index];
+    }
+
+    return nullptr;
+}
+
+CHIP_ERROR AccessRestriction::CreateCommissioningEntry(SharedPtr<Entry> entry)
+{
+    if (!entry->IsValid())
+    {
+        ChipLogError(DataManagement, "AccessRestriction: invalid restriction entry");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    sCommissioningEntries.push_back(entry);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessRestriction::CreateEntry(size_t * index, const Entry & entry, FabricIndex fabricIndex)
+{
+    if (!entry.IsValid())
+    {
+        ChipLogError(DataManagement, "AccessRestriction: invalid restriction entry");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    auto localEntry = MakeShared<Entry>(entry);
+    size_t newIndex;
+
+    localEntry->fabricIndex = fabricIndex;
+
+    if (mFabricEntries.find(fabricIndex) == mFabricEntries.end())
+    {
+        mFabricEntries[fabricIndex] = std::vector<SharedPtr<Entry>>();
+        mFabricEntries[fabricIndex].push_back(localEntry);
+    }
+    else
+    {
+        mFabricEntries[fabricIndex].push_back(localEntry);
+    }
+
+    newIndex = mFabricEntries[fabricIndex].size() - 1;
+
+    for (EntryListener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+    {
+        listener->OnEntryChanged(fabricIndex, newIndex, localEntry, EntryListener::ChangeType::kAdded);
+    }
+
+    if (index != nullptr)
+    {
+        *index = newIndex;
+    }
+
+    ChipLogProgress(DataManagement, "AccessRestriction: update entry f=%u i=%lu", fabricIndex, newIndex);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessRestriction::DeleteEntry(size_t index, const FabricIndex fabricIndex)
```

Comment:
Is the idea here that you can delete entries base on the returned index value from the add? It seems like that would fall apart as soon as anything was deleted.

Can this function signature be changed to not use the index (and above not to return it)? Index is not a stable identifier in a vector.

Reply:
The data management model has been greatly simplified and this is now gone.